contracts/telospooldex/telospooldex.cpp

```diff
@@ -0,0 +1,297 @@
+                    row.currency_reserve += fund_it->currency;
+                });
+
+                funding_attempts.erase(fund_it);
+            }
+            return;
+        }
+
+        case PROTO_VERSION.value : {  
+            // general protocol parsing
+            asset min = asset_from_string(memo_tokens[2]);
+            name recipient = name(memo_tokens[3]);
+            check(is_account(recipient), ERR_RECIPIENT_NOT_FOUND);
+            
+            // get first element of path
+            string path_str = memo_tokens[1];
+            vector<string> jumps = split(path_str, " ");
+            check(jumps.size() > 0, ERR_EMPTY_PATH);
+
+            vector<string> conversion_data = split(jumps.front(), "/");
+            check(conversion_data.size() == 2, ERR_MEMO_PARSING);
+            check(name(conversion_data[0]) == get_self(), ERR_INCORRECT_CONVERTER);
+
+            // find pool
+            symbol_code A = quantity.symbol.code();
+            symbol_code B = symbol_code(conversion_data[1]);
+
+            pools pool_markets(get_self(), get_self().value);
+            auto sym_index = pool_markets.get_index<"symbols"_n>();
+            auto pool_it = sym_index.find(symbols_get_index(A, B).value);
+
+            if (pool_it == sym_index.end())
+                pool_it = sym_index.find(symbols_get_index(B, A).value);
+
+            check(pool_it != sym_index.end(), ERR_POOL_NOT_FOUND);
+
+            // check if pool has funds
+            check(pool_it->commodity_reserve.amount > 0, ERR_POOL_NOT_FUNDED);
+            check(pool_it->currency_reserve.amount > 0, ERR_POOL_NOT_FUNDED);
+
+            // make conversion
+            tuple<asset, asset> result = get_conversion(pool_it->id, quantity);
+            asset total = get<0>(result);
+            asset rate = get<1>(result);
+  
+            // update pool reserves
+            sym_index.modify(pool_it, get_self(), [&](auto& row) {
+                if (quantity.symbol.code() == row.commodity_reserve.symbol.code()) {
+                    row.commodity_reserve += quantity;
+                    row.currency_reserve -= total;
+                } else {
+                    row.currency_reserve += quantity;
+                    row.commodity_reserve -= total;
+                }
+            });
+
+            // pop first
+            jumps.erase(jumps.begin());
+
+            if (jumps.size() == 0) {
+                // last jump of path, send tokens to recipient
+
+                print("total: ", total, '\n');
+                print("rate: ", rate, '\n');
+                check(total >= min, ERR_BAD_DEAL);
+               
+                // TODO: final transaction memo
+                action(
+                    permission_level{get_self(), "active"_n},
+                    get_contract_for_token(total.symbol.code()),
+                    "transfer"_n, 
+                    make_tuple(
+                        get_self(), recipient, total, string(""))
```

Comment:
When sending the final tokens to the end-user after conversion, a "thank you" memo should be included for using the telos pool DEX. Something like: "Thank you for using Telos Pool DEX"